crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+    /// The factor used to determine the size of the time slices.
+    ///
+    /// The size of a time slice is 2^factor * [UNIT_TIME].
+    /// Partial slices are created for "recent time" and are created in decreasing sizes from
+    /// 2^factor * [UNIT_TIME] down to 2^0 * [UNIT_TIME].
+    factor: u8,
+    /// The number of full time slices that have been stored.
+    ///
+    /// These full slices are always of size 2^factor * [UNIT_TIME].
+    full_slices: u64,
+    /// The partial slices, with hashes stored in memory.
+    ///
+    /// Because these change every [UNIT_TIME], they are stored in memory.
+    partial_slices: Vec<PartialSlice>,
+    /// The minimum amount of time that must be reserved for recent time.
+    ///
+    /// The algorithm is free to allocate up to 2x this value as recent time. This allows for the
+    /// algorithm to create more slices that fill more of the recent time window.
+    min_recent_time: Duration,
+    /// The timestamp at which the next update is required.
+    ///
+    /// After this time, there will be an excess amount of time that could be allocated into a
+    /// partial slice. The data structure is still usable but will get behind if
+    /// [PartitionedTime::update] is not called.
+    /// It is idempotent to call [PartitionedTime::update] more often than required, but it is not
+    /// efficient.
+    next_update_at: Timestamp,
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct PartialSlice {
+    /// The start timestamp of the time slice.
+    ///
+    /// This timestamp is included in the time slice.
+    start: Timestamp,
+    /// Size is used in the formula 2^size * [UNIT_TIME].
+    ///
+    /// That means a 0 size translates to [UNIT_TIME], and a 1 size translates to 2*[UNIT_TIME].
+    size: u8,
+    /// The combined hash over all the ops in this time slice.
+    hash: bytes::Bytes,
+}
+
+impl PartialSlice {
+    /// The end timestamp of the time slice.
+    ///
+    /// This timestamp is not included in the time slice.
+    fn end(&self) -> Timestamp {
+        self.start
+            + Duration::from_secs((1u64 << self.size) * UNIT_TIME.as_secs())
+    }
+}
+
+// Public methods
+impl PartitionedTime {
+    /// Create a new instance of [PartitionedTime] from the given store.
+    ///
+    /// The store is needed to check how many time slices were created last time this
+    /// [PartitionedTime] was updated, if any.
+    ///
+    /// The method will then update the state of the [PartitionedTime] to the current time.
+    /// It does this by checking that the construction of this [PartitionedTime] is consistent
+    /// and then calling [PartitionedTime::update].
+    ///
+    /// The resulting [PartitionedTime] will be consistent with the store and the current time.
+    /// It should be updated again after [PartitionedTime::next_update_at].
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if factor >= 54.
```

Comment:
Why does it panic?

Reply:
Integer overflow, the numbers just get too large. 

This is a good point though, because if we let this be settable by apps then maybe we need to have this return an error and reject the app install rather than crashing the whole conductor... Yeah I should change this I think